go/store/prolly/sort/external.go

```diff
@@ -0,0 +1,458 @@
+
+	newF, err := a.newFile()
+	if err != nil {
+		return err
+	}
+	outF := newKeyFile(newF, a.batchSize)
+	m, err := newFileMerger(ctx, a.keyCmp, outF, fileLevel[:a.fileMax]...)
+	if err != nil {
+		return err
+	}
+	if err := m.run(ctx); err != nil {
+		return err
+	}
+
+	// zero out compacted level
+	// add to next level
+	a.files[level] = a.files[level][a.fileMax:]
+	if len(a.files) <= level+1 {
+		a.files = append(a.files, []keyIterable{outF})
+	} else {
+		a.files[level+1] = append(a.files[level+1], outF)
+	}
+
+	return nil
+}
+
+func newKeyMem(size int) *keyMem {
+	return &keyMem{byteLim: size}
+}
+
+type keyMem struct {
+	keys    []val.Tuple
+	byteCnt int
+	byteLim int
+}
+
+func (k *keyMem) insert(key val.Tuple) bool {
+	if len(key)+int(keyLenSz)+k.byteCnt > k.byteLim {
+		return false
+	}
+
+	k.keys = append(k.keys, key)
+	k.byteCnt += len(key) + int(keyLenSz)
+	return true
+}
+
+func (k *keyMem) flush(f *os.File, cmp func(val.Tuple, val.Tuple) bool) (*keyFile, error) {
+	k.sort(cmp)
+	kf := newKeyFile(f, k.byteLim)
+	for _, k := range k.keys {
+		if err := kf.append(k); err != nil {
+			return nil, err
+		}
+	}
+	if err := kf.buf.Flush(); err != nil {
+		return nil, err
+	}
+	return kf, nil
+}
+
+func (f *keyMem) IterAll(_ context.Context) (keyIter, error) {
+	return &keyMemIter{keys: f.keys}, nil
+}
+
+type keyMemIter struct {
+	keys []val.Tuple
+	i    int
+}
+
+func (i *keyMemIter) Next(ctx context.Context) (val.Tuple, error) {
+	if i.i >= len(i.keys) {
+		return nil, io.EOF
+	}
+	ret := i.keys[i.i]
+	i.i++
+	return ret, nil
+}
+
+func (i *keyMemIter) Close() {
+	return
+}
+
+func (k *keyMem) Len() int {
+	return len(k.keys)
+}
+
+// sort sorts the tuples in memory without flushing to disk
+func (k *keyMem) sort(cmp func(val.Tuple, val.Tuple) bool) {
+	sort.Slice(k.keys, func(i, j int) bool {
+		return cmp(k.keys[i], k.keys[j])
+	})
+}
+
+func newKeyFile(f *os.File, batchSize int) *keyFile {
+	return &keyFile{f: f, buf: bufio.NewWriterSize(f, batchSize), batchSize: batchSize}
+}
+
+type keyFile struct {
+	f         *os.File
+	buf       *bufio.Writer
+	batchSize int
+}
+
+func (f *keyFile) IterAll(ctx context.Context) (keyIter, error) {
+	if f.batchSize == 0 {
+		return nil, fmt.Errorf("invalid zero batch size")
+	}
+	if _, err := f.f.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+	return &keyFileReader{buf: bufio.NewReader(f.f)}, nil
+}
+
+var keySizePool = sync.Pool{
+	// New is called when a new instance is needed
+	New: func() interface{} {
+		return make([]byte, keyLenSz)
+	},
+}
+
+// GetKeySizeBuf fetches a buffer from the pool
+func GetKeySizeBuf() []byte {
+	return keySizePool.Get().([]byte)
+}
+
+// PutKeySizeBuf returns a buffer to the pool
+func PutKeySizeBuf(b []byte) {
+	keySizePool.Put(b)
+}
+
+// append writes |keySize|key| to the intermediate file
+func (f *keyFile) append(k val.Tuple) error {
+	v := uint32(len(k))
+	sizeBuf := GetKeySizeBuf()
+	defer PutKeySizeBuf(sizeBuf)
+	writeUint32(sizeBuf, v)
```

Comment:
keySizePool and friends here look wrong to me. You can just use an array on the stack.

```go
var sizeBuf [4]byte
writeUint32(sizeBuf[:], v)
if _, err := f.buf.Write(sizeBuf[:]); err != nil { 
    ....
```